### PR TITLE
Handshake with FW version and clock, Notify with priority

### PR DESCRIPTION
**Pull request by marcos-diaz**
#39 

Screenshot 2024-05-16 at 19 17 08

### CODE REVIEW

File: src/components/header/header.ts [hunk 1]
```diff
@@ -78,8 +82,17 @@ export class HeaderComponent {
     localStorage.setItem(FIRMWARE_ACK, fwValue)
   }
 
+  shouldWarningFirmware() {
```
Comment (przymi) [inline src/components/header/header.ts:85]: "should" => "display" or better "requestDisplay"

File: src/components/header/header.ts [hunk 2]
```diff
@@ -78,8 +82,17 @@ export class HeaderComponent {
...
+    const latestVersion = this.firmwareAsNumber(LATEST_FIRMWARE)
+    const ackVersion = Number(localStorage.getItem(FIRMWARE_ACK))
+    return ackVersion < latestVersion
+  }
+
   shouldNotifyFirmware() {
-    const knownByUser = Number(localStorage.getItem(FIRMWARE_ACK))
-    return knownByUser < this.firmwareAsNumber(LATEST_FIRMWARE)
+    if (!this.webusb.isConnectedRaw) return false
+    const latestVersion = this.firmwareAsNumber(LATEST_FIRMWARE)
```
Comment (przymi) [inline src/components/header/header.ts]: In ctrl/src/lib/version.ts rename 
```LATEST_FIRMWARE``` to ```LATEST_FIRMWARE_VERSION```

File: src/lib/ctrl.ts [hunk 3]
```diff
@@ -621,4 +625,51 @@ export class CtrlGyroAxis extends CtrlSection {
...
+export class CtrlHandshakeGet extends Ctrl {
+  constructor(
+  ) {
+    super(1, DeviceId.ALPAKKA, MessageType.HANDSHAKE_GET)
+  }
+}
+
+export class CtrlHandshakeSet extends Ctrl {
+  constructor(
+    public time: number
```
Comment (przymi) [inline src/lib/ctrl.ts:638]: Why not some datetime type?

File: src/lib/ctrl.ts [hunk 4]
```diff
@@ -621,4 +625,51 @@ export class CtrlGyroAxis extends CtrlSection {
...
+  ) {
+    super(1, DeviceId.ALPAKKA, MessageType.HANDSHAKE_SET)
+  }
+
+  override payload() {
```
Comment (przymi) [inline src/lib/ctrl.ts:643]: Find a better way

File: src/lib/ctrl.ts [hunk 5]
```diff
@@ -621,4 +625,51 @@ export class CtrlGyroAxis extends CtrlSection {
   }
 }
 
+export class CtrlHandshakeGet extends Ctrl {
+  constructor(
+  ) {
+    super(1, DeviceId.ALPAKKA, MessageType.HANDSHAKE_GET)
```
Comment (przymi) [inline src/lib/ctrl.ts]: Make this a constant, as it is re-used in at least two more places below.

File: src/lib/ctrl.ts [hunk 6]
```diff
@@ -621,4 +625,51 @@ export class CtrlGyroAxis extends CtrlSection {
...
+      dataview.getUint8(7),
+    ]
+  }
+}
+
+export class CtrlHandshakeShare extends Ctrl {
+  constructor(
+    public version: number[],
+  ) {
+    super(1, DeviceId.ALPAKKA, MessageType.HANDSHAKE_SHARE)
```
Comment (przymi) [inline src/lib/ctrl.ts]: Constant, see above

File: src/lib/ctrl.ts [hunk 7]
```diff
@@ -621,4 +625,51 @@ export class CtrlGyroAxis extends CtrlSection {
...
+  ) {
+    super(1, DeviceId.ALPAKKA, MessageType.HANDSHAKE_GET)
+  }
+}
+
+export class CtrlHandshakeSet extends Ctrl {
+  constructor(
+    public time: number
+  ) {
+    super(1, DeviceId.ALPAKKA, MessageType.HANDSHAKE_SET)
```
Comment (przymi) [inline src/lib/ctrl.ts]: Constant, see above

File: src/lib/ctrl.ts [hunk 8]
```diff
@@ -621,4 +625,51 @@ export class CtrlGyroAxis extends CtrlSection {
...
+  constructor(
+    public version: number[],
+  ) {
+    super(1, DeviceId.ALPAKKA, MessageType.HANDSHAKE_SHARE)
+  }
+
+  static override decode(buffer: ArrayBuffer) {
+    const data = Array.from(new Uint8Array(buffer))
+    const version: number[] = []
+    version[0] = data[4]
```
Comment (przymi) [inline src/lib/ctrl.ts]: What is data[0 - 3]

File: src/services/webusb.ts [hunk 9]
```diff
@@ -37,6 +40,7 @@ interface PresetWithValues {
 export class WebusbService {
   browserIsCompatible = false
   device: any = null
+  deviceVersion: number[] = [0, 0, 0]
```
Comment (przymi) [inline src/services/webusb.ts]: remove ```number[]```

**Review by przymi — CHANGES_REQUESTED** [review]

File: src/components/header/header.ts (hunk 2 above)
Comment (marcos-diaz) [reply src/components/header/header.ts]: Renamed to `MINUMUM_FIRMWARE_VERSION`

File: src/lib/ctrl.ts (hunk 3 above)
Comment (marcos-diaz) [reply src/lib/ctrl.ts:638]: Coming in ES2025 [https://github.com/tc39/proposal-temporal](https://github.com/tc39/proposal-temporal)

**Pull request merged**